src/sentry/utils/arroyo_router.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+import time
+from collections import deque
+from collections.abc import Callable, Mapping, MutableMapping, Sequence
+from typing import Any, Deque, Generic, TypeVar
+
+from arroyo.processing.strategies import MessageRejected
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Message, Partition, TStrategyPayload
+
+from sentry.utils.arroyo_guard import NonFinalStrategy, guard
+
+TResult = TypeVar("TResult")
+
+
+class MergeRoutes(ProcessingStrategy[TResult], Generic[TResult]):
+    def __init__(self, routing_key: str, buffer: MessageBuffer[TResult]) -> None:
+        self.routing_key = routing_key
+        self.buffer = buffer
+
+    def submit(self, message: Message[TResult]) -> None:
+        self.buffer.add(message, self.routing_key)
+
+    def poll(self) -> None:
+        pass
+
+    def join(self, timeout: float | None = None) -> None:
+        pass
+
+    def close(self) -> None:
+        pass
+
+    def terminate(self) -> None:
+        pass
+
+
+@guard
+class Router(
+    NonFinalStrategy[TStrategyPayload, TResult],
+):
+    def __init__(
+        self,
+        next_step: ProcessingStrategy[TResult],
+        route_builders: Mapping[
+            str, Callable[[MergeRoutes[TResult]], ProcessingStrategy[TStrategyPayload]]
+        ],
+        routing_func: Callable[[Message[TStrategyPayload]], str],
+    ) -> None:
+        self.buffer: MessageBuffer[TResult] = MessageBuffer(list(route_builders.keys()))
+        self.routing_func = routing_func
+        self.next_step = next_step
+
+        self.routes: Mapping[str, ProcessingStrategy[TStrategyPayload]] = {
+            routing_key: builder(MergeRoutes(routing_key, self.buffer))
+            for (routing_key, builder) in route_builders.items()
+        }
+
+    def submit(self, message: Message[TStrategyPayload]) -> None:
+        if len(self.buffer) > 10000:
+            raise MessageRejected
+
+        message_route = self.routing_func(message)
+        self.routes[message_route].submit(message)
+        self.buffer.remove(message, message_route)
```

Comment:
I am not sure I follow this process in the following cases:
- Messages filtered by one of the routes. We would try to remove a message that potentially was never added to the buffer. Is that ok ? 
- Batching messages. Let's say one route is baching messages and the other is not. If we send 10 messages to the batching strategy then a message to the non batching one then 10 messages to the batching strategy, we may remove the individual message before the batch is even added. Is this a problem ?

Reply:
we should probably change the remove api but it does nothing with the message itself, so filtered or batched or whatever doesn't matter. it just marks those offsets as complete for that route.